src/dir_contents.ml

```diff
@@ -374,246 +193,461 @@ let build_coq_modules_map (d : _ Dir_with_dune.t) ~dir ~modules =
       Lib_name.Map.add map (Dune_file.Coq.best_name coq) modules
     | _ -> map)
 
-type result0_here = {
-  t : t;
-  (* [rules] includes rules for subdirectories too *)
-  rules : Rules.t option;
-  subdirs : t Path.Build.Map.t;
-}
+module rec Load : sig
+  val get : Super_context.t -> dir:Path.Build.t -> t
+  val gen_rules : Super_context.t -> dir:Path.Build.t -> gen_rules_result
+end = struct
+  let virtual_modules sctx vlib =
+    let info = Lib.info vlib in
+    let lib_modules =
+      match Option.value_exn (Lib_info.virtual_ info) with
+      | External lib_modules -> lib_modules
+      | Local ->
+        let src_dir =
+          Lib_info.src_dir info
+          |> Path.as_in_build_dir_exn
+        in
+        let t = Load.get sctx ~dir:src_dir in
+        modules_of_library t ~name:(Lib.name vlib)
+    in
+    let existing_virtual_modules =
+      Lib_modules.virtual_modules lib_modules
+      |> Module.Name.Map.keys
+      |> Module.Name.Set.of_list
+    in
+    let allow_new_public_modules =
+      Lib_modules.wrapped lib_modules
+      |> Wrapped.to_bool
+      |> not
+    in
+    { Modules_field_evaluator.Implementation.
+      existing_virtual_modules
+    ; allow_new_public_modules
+    }
 
-type result0 =
-  | See_above of Path.Build.t
-  | Here of result0_here
+  let make_modules sctx (d : _ Dir_with_dune.t) ~modules =
+    let scope = d.scope in
+    let libs, exes =
+      List.filter_partition_map d.data ~f:(fun stanza ->
+        match (stanza : Stanza.t) with
+        | Library lib ->
+          let src_dir = d.ctx_dir in
+          let resolved = lazy (
+            let name = Library.best_name lib in
+            Lib.DB.find_even_when_hidden (Scope.libs scope) name
+            (* can't happen because this library is defined using the current
+               stanza *)
+            |> Option.value_exn
+          ) in
+          let kind : Modules_field_evaluator.kind =
+            match lib.implements, lib.virtual_modules with
+            | Some _, None ->
+              (* diml: this [Result.ok_exn] means that if the user
+                 writes an invalid [implements] field, we will get an
+                 error immediately even if the library is not
+                 built. We should change this to carry the [Or_exn.t]
```

Comment:
Indeed. I believe this issue can be fixed once we have a "raw" library database that just resolves names. Do we have a ticket for this somewhere yet?

Reply:
I don't believe so, it's worth adding one indeed